### PR TITLE
Hydroelectric Generator Introduction

### PR DESCRIPTION
**Pull request by MSieghardt**
The way it works:
Water goes into basin.
Basin fills up.
release water to generate power.

## About The Pull Request
<details>
<summary>
	A new hydroelectric generator dam.

We have a console that controls the water and turbines that generate the energy by draining the water. There's a nice sheet of paper that explains how it works. 

Tested on my server extensively.
</summary>
<hr>

	
<hr>
</details>

## Changelog
:cl:

/:cl:

### CODE REVIEW

**Comment by Kazkin** [comment]
This looks pretty good, not to mention the code appears pretty clean. I'll let @KathyRyals take a look before I approve it in case they see something I missed.

File: code/modules/power/hydroelectric.dm [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+		return
+	if(malfstate != 0)
+		add_overlay(image('icons/obj/machines/thermoelectric.dmi', icon_state = "circ-hot", layer = FLY_LAYER))
+	else if (control.working == 1)
+		add_overlay(image('icons/obj/machines/thermoelectric.dmi', icon_state = "circ-run", layer = FLY_LAYER))
+	return
+
+/obj/machinery/power/hydroelectric/proc/Malfunction()
+	malfstate = rand(1,2)
+	if(malfstate == 1)
```
Comment (KathyRyals) [inline code/modules/power/hydroelectric.dm:36]: Need to use bitflags instead of random values.

**Review by KathyRyals — CHANGES_REQUESTED** [review]

File: code/modules/power/hydroelectric.dm [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+	var/obj/machinery/power/hydroelectric_control/control = null
+
+/obj/machinery/power/hydroelectric/drain_power()
+	return -1
+
+/obj/machinery/power/hydroelectric/update_icon()
+	cut_overlays()
+	if(!control)
+		return
+	if(malfstate != 0)
```
[the remark this answers is not in the document]
Comment (Trilbyspaceclone) [reply code/modules/power/hydroelectric.dm:28]: wouldnt this be if(malfstate) ?
atm its "If malfstate not falsey" so double negative? - im asking idk and wish to learn on this

File: code/modules/power/hydroelectric.dm (hunk 1 above)
Comment (MSieghardt) [reply code/modules/power/hydroelectric.dm:36]: The idea is that we'd have a random broken state. 1 and 2 are different broken states, with each requiring its own method of fixing.

File: code/modules/power/hydroelectric.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (KathyRyals) [reply code/modules/power/hydroelectric.dm:28]: Yeah, my bad. Typed too fast.

File: code/modules/power/hydroelectric.dm (hunk 1 above)
Comment (KathyRyals) [reply code/modules/power/hydroelectric.dm:36]: Yeah, I know. Use bitflags instead of a cryptic 1 and 2.

Comment (KathyRyals) [reply code/modules/power/hydroelectric.dm:36]: Or, if you really can't do bitflags, use constants.

**Comment by MSieghardt** [comment]
Changed it.

File: code/modules/power/hydroelectric.dm [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+	name = "hydroelectric turbine"
+	desc = "A hydroelectric generator able to store water and then release it to generate electricity. This one seems to be important to the structural integrity around and is attached directly to the dam's frame."
+	icon = 'icons/obj/machines/thermoelectric.dmi'
+	icon_state = "circ-unassembled"
+	density = 1
+	anchored = 1
+	use_power = NO_POWER_USE
+	idle_power_usage = 0
+	active_power_usage = 0
+	health = 10
```
Comment (Trilbyspaceclone) [inline code/modules/power/hydroelectric.dm:17]: Why is the hp so low? I can imagine theirs cases were an engi hits the turbine by misstake with their crowbar and kills it or something... Maybe 300?

**Comment by Kazkin** [comment]
Going to wait for Lessie to approve the changes to ensure they work.

**Comment by MSieghardt** [comment]
I tested it after changing things too, should still work.

**Review by KathyRyals — APPROVED** [review]

**Pull request merged**